3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/SendgridAdapter.scala

```diff
@@ -69,53 +52,93 @@ object SendgridAdapter extends Adapter {
+    "processed" -> SchemaKey("com.sendgrid", "processed", "jsonschema", "1-0-0").toSchemaUri,
+    "dropped" -> SchemaKey("com.sendgrid", "dropped", "jsonschema", "1-0-0").toSchemaUri,
+    "delivered" -> SchemaKey("com.sendgrid", "delivered", "jsonschema", "1-0-0").toSchemaUri,
+    "deferred" -> SchemaKey("com.sendgrid", "deferred", "jsonschema", "1-0-0").toSchemaUri,
+    "bounce" -> SchemaKey("com.sendgrid", "bounce", "jsonschema", "1-0-0").toSchemaUri,
+    "open" -> SchemaKey("com.sendgrid", "open", "jsonschema", "1-0-0").toSchemaUri,
+    "click" -> SchemaKey("com.sendgrid", "click", "jsonschema", "1-0-0").toSchemaUri,
+    "spamreport" -> SchemaKey("com.sendgrid", "spamreport", "jsonschema", "1-0-0").toSchemaUri,
+    "unsubscribe" -> SchemaKey("com.sendgrid", "unsubscribe", "jsonschema", "1-0-0").toSchemaUri,
+    "group_unsubscribe" -> SchemaKey("com.sendgrid", "group_unsubscribe", "jsonschema", "1-0-0").toSchemaUri,
+    "group_resubscribe" -> SchemaKey("com.sendgrid", "group_resubscribe", "jsonschema", "1-0-0").toSchemaUri
   )
 
+
   /**
-   * Converts a CollectorPayload instance into raw events.
-   * A Sendgrid Tracking payload only contains a single event.
-   * We expect the name parameter to be 1 of 6 options otherwise
-   * we have an unsupported event type.
-   *
-   * @param payload The CollectorPayload containing one or more
-   *        raw events as collected by a Snowplow collector
-   * @param resolver (implicit) The Iglu resolver used for
-   *        schema lookup and validation. Not used
-   * @return a Validation boxing either a NEL of RawEvents on
-   *         Success, or a NEL of Failure Strings
-   */
+    *
+    * Converts a payload into a list of validated events
+    * Expects a valid json - returns a single failure if one is not present
+    *
+    * @param body json payload as POST'd by sendgrid
+    * @param payload the rest of the payload details
+    * @return a list of validated events, successes will be the corresponding raw events
+    *         failures will contain a non empty list of the reason(s) for the particular event failing
+    */
+  private def payloadBodyToEvents(body: String, payload: CollectorPayload): List[Validated[RawEvent]] = {
+    try {
+      val parsed = parse(body)
+
+      if (parsed.children.isEmpty) {
+        return List(s"$VendorName event failed json sanity check: has no events".failNel)
+      }
+
+      for ((itm, index) <- parsed.children.zipWithIndex)
+        yield {
+          val eventType = (itm \\ "event").extractOpt[String]
+          val queryString = toMap(payload.querystring)
+
+          lookupSchema(eventType, VendorName, index, EventSchemaMap) match {
+            case Success(schema) => {
+              Success(
+                RawEvent(
+                  api = payload.api,
+                  parameters = toUnstructEventParams(TrackerVersion,
+                    queryString,
+                    schema,
+                    itm,
```

Comment:
With the final payload that you send in there are a few things we can manipulate beforehand:

* The `event` key-pair can be removed as it is extra data for no gain.
* The `timestamp` fields can be formatted from UNIX Epoch to something a bit more readable.
  - `1415709559` -> `2014-11-11T12:39:19.000Z`

@alexanderdean can you think of any other manipulations that could be useful?